### PR TITLE
IBP-3107 Move Experimental Design to BMSAPI

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/java/impl/middleware/design/generator/ExperimentDesignGenerator.java [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+
+	@Resource
+	private OntologyDataManager ontologyDataManager;
+
+	@Resource
+	private StandardVariableTransformer standardVariableTransformer;
+
+	private final Random random = new Random();
+
+	public MainDesign createRandomizedCompleteBlockDesign(
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/design/generator/ExperimentDesignGenerator.java]: Kind of weird how this is implemented. In only one class we do have createXXXDesign without following any rule and not taking advantage of polymorphism as we have in another areas of experimental design generation. Something we can improve

File: src/main/java/org/ibp/api/java/impl/middleware/design/transformer/StandardVariableTransformer.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.generationcp.middleware.domain.dms.StandardVariable;
+import org.generationcp.middleware.domain.etl.MeasurementVariable;
+import org.generationcp.middleware.domain.ontology.VariableType;
+import org.springframework.stereotype.Component;
+
+@Component
+public class StandardVariableTransformer {
+
+	public MeasurementVariable convert(final StandardVariable standardVariable, final VariableType variableType) {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/design/transformer/StandardVariableTransformer.java]: This code seems familiar to me, is it not repeated in other places in the application?

File: src/main/java/org/ibp/api/java/impl/middleware/design/util/ExpDesignUtil.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import org.ibp.api.rest.dataset.ObservationUnitData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Marshaller;
+import java.io.StringWriter;
+
+public class ExpDesignUtil {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/java/impl/middleware/design/util/ExpDesignUtil.java]: ExperimentalDesignUtil sounds better

File: src/main/java/org/ibp/api/rest/design/BVDesignProperties.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package org.ibp.api.rest.design;
+
+import org.springframework.core.env.Environment;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.Resource;
+
+@Component
+public class BVDesignProperties {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/rest/design/BVDesignProperties.java]: java classes names should be singular by convention. Let's try to use BVDesignPropertiesConfig or something similar

File: src/main/java/org/ibp/api/domain/design/License.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package org.ibp.api.domain.design;
+
+public class License
+{
+	private String message;
+
+	private String expiryDays;
+
+	private String expiry;
```
[the remark this answers is not in the document]
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/domain/design/License.java]: Expiry here is a string representation of date when the BV license will expire. 
A sample JSON output when running BVDesign -status -json is:
`{"status":{"appStatus":"Successful license checkout","returnCode":0,"license":{"message":"Successful license checkout","expiry":"31-dec-2019","expiryDays":0}}}`
Hence the naming of the classes/properties.

File: src/main/java/org/ibp/api/domain/design/DesignLicenseInfo.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package org.ibp.api.domain.design;
+
+public class DesignLicenseInfo {
```
[the remark this answers is not in the document]
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/domain/design/DesignLicenseInfo.java]: This is used by BVDesign but we wanted to keep the naming generic

File: src/main/java/org/ibp/api/java/impl/middleware/design/generator/ExperimentDesignGenerator.java (hunk 1 above)
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/java/impl/middleware/design/generator/ExperimentDesignGenerator.java]: Yes I agree, will create a tech backlog ticket for this.

File: src/main/java/org/ibp/api/java/impl/middleware/design/transformer/StandardVariableTransformer.java (hunk 2 above)
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/java/impl/middleware/design/transformer/StandardVariableTransformer.java]: Removed **StandardVariableTransformer** class and overloaded **MeasurementVariableTransformer** in MIddleware for use in this module

File: src/main/java/org/ibp/api/rest/design/BVDesignProperties.java (hunk 4 above)
Comment (darla-leafnode) [reply src/main/java/org/ibp/api/rest/design/BVDesignProperties.java]: Was thinking of cleaning this up today too. I ended up just deleting this class and resolving the individual properties where they are needed

File: src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java [hunk 7]
```diff
@@ -17,23 +17,28 @@
...
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-@Api(value = "Breeding View", description = "Web Services to process the Breeding View output")
+@Api(value = "Breeding View Resource")
 @RestController
 @RequestMapping("/breeding_view")
 public class BreedingView {
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java]: Class name is wrong, by convention we are appending Resource at the end of the class

File: src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java [hunk 8]
```diff
@@ -17,23 +17,28 @@
...
+import javax.annotation.Resource;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
-@Api(value = "Breeding View", description = "Web Services to process the Breeding View output")
+@Api(value = "Breeding View Resource")
 @RestController
 @RequestMapping("/breeding_view")
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java]: can we define breeding_view as a resource itself? it the answer is YES, then it should be called breeding-views instead.

File: src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java [hunk 9]
```diff
@@ -102,16 +112,27 @@ public DataResponse saveSsaResultSummary(
...
+		} catch (final Exception e) {
 			BreedingView.LOG.debug(e.getMessage(), e);
 			response = new DataResponse(false, "Failed to invoke service: " + e.toString());
 		}
 
 		return response;
 	}
 
-	public void setBreedingViewService(BreedingViewService breedingViewService) {
+	@ApiOperation(value = "Count number of days before Breeding View license expires")
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java]: Please move this into a new BVLicenseResource, this class does not make much sense, and will have to be reviewed later

File: src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java [hunk 10]
```diff
@@ -43,31 +48,36 @@
 	@Autowired
 	private BreedingViewService breedingViewService;
 
+	@Resource
+	private DesignLicenseService designLicenseService;
+
 	@ApiOperation(value = "Save the Single-Site Analysis CSV output file with heritability", notes = "", response = DataResponse.class)
 	@RequestMapping(value = "{cropName}/ssa/save_result_summary", method = RequestMethod.GET, produces = {MediaType.APPLICATION_XML_VALUE})
 	@ResponseBody
 	public DataResponse saveSsaResultSummary(
```
Comment (clarysabel) [inline src/main/java/org/ibp/api/ibpworkbench/rest/BreedingView.java]: This resource and class does not follow any REST standard at all 👎 We should review it later, but I left a comment to create a different resource for the BVLicense

File: src/main/java/org/ibp/api/rest/design/ExperimentDesignResource.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+@Api(value = "Experiment Design Service")
+@Controller
+@RequestMapping("/crops")
+public class ExperimentDesignResource {
+
+	@Resource
+	private ExperimentDesignService experimentDesignService;
+
+	@ApiOperation(value = "Generate experiment design", notes = "Generate experiment design")
+	@RequestMapping(value = "/{crop}/studies/{studyId}/design", method = RequestMethod.POST)
```
[the remark this answers is not in the document]
Comment (clarysabel) [reply src/main/java/org/ibp/api/rest/design/ExperimentDesignResource.java]: @darla-leafnode  please check this comment before merging

File: src/main/java/org/ibp/api/rest/design/ExperimentDesignResource.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+		@RequestBody final ExperimentDesignInput experimentDesignInput) {
+
+		this.experimentDesignService.generateAndSaveDesign(crop, studyId, experimentDesignInput);
+
+		return new ResponseEntity<>(HttpStatus.OK);
+	}
+
+
+	@ApiOperation(value = "Delete experiment design", notes = "Delete experiment design")
+	@RequestMapping(value = "/{crop}/studies/{studyId}/design", method = RequestMethod.DELETE)
```
[the remark this answers is not in the document]
Comment (clarysabel) [reply src/main/java/org/ibp/api/rest/design/ExperimentDesignResource.java]: @darla-leafnode please check this comment before merging

**Pull request merged**